manifests/database/default_read_grant.pp

```diff
@@ -19,7 +19,7 @@
                   acl.defaclacl
                 FROM pg_default_acl acl
                 JOIN pg_namespace ns ON acl.defaclnamespace=ns.oid
-                WHERE acl.defaclacl::text ~ '.*\\\\\"${database_read_only_username}\\\\\"=r/${database_username}\\\".*'
+                WHERE array_to_string(acl.defaclacl, ' + ') ~ '.*(\")?${database_read_only_username}(\")?=r/(\")?${database_username}(\")?.*'
```

Comment:
The multiple layers of escaping make the original code barely readable.  Your proposed change looks better :+1:

A few notes to improve this:

1. Regexp are not anchored, so this is equivalent to:
```suggestion
                WHERE array_to_string(acl.defaclacl, ' + ') ~ '(\")?${database_read_only_username}(\")?=r/(\")?${database_username}(\")?'
```

2. The optional quotes work in pairs, so if an opening quote is present, we must match a closing one. We can do this with captures:

```suggestion
                WHERE array_to_string(acl.defaclacl, ' + ') ~ '(\"?)${database_read_only_username}\\1=r/(\"?)${database_username}\\2'
```

3. Since we can't match on array items, joining the items is fine but we should put separators before the first item and after the last so that we can ensure we are matching a complete entry:

```suggestion
                WHERE ' + ' || array_to_string(acl.defaclacl, ' + ') || ' + ' ~ ' + (\"?)${database_read_only_username}\\1=r/(\"?)${database_username}\\2 + '
```

Reply:
The second suggestion works. 

The last suggestion doesn't work. The query returns 0 rows, and as such the ALTER table is run each time puppet runs.

FWIW, I've tried to dump some output from psql so it may be clearer.

```
postgres@puppetdb:~$ psql -d puppetdb -t -X -c "SELECT acl.defaclacl FROM pg_default_acl acl"
 {"\"puppetdb-read\"=U/puppetdb"}
 {"\"puppetdb-read\"=X/puppetdb"}
 {"\"puppetdb-read\"=r/puppetdb"}

postgres@puppetdb:~$ psql -d puppetdb -t -X -c "SELECT array_to_string(acl.defaclacl, ' + ') FROM pg_default_acl acl"
 "puppetdb-read"=U/puppetdb
 "puppetdb-read"=X/puppetdb
 "puppetdb-read"=r/puppetdb
```

Reply:
Hum, this does not work because `+` has a special meaning in a regexp.

Using another non-special char may be better, such as `@`.